# DTSC 580: Data Manipulation
## Assignment: LEGO Project
### Name: 

![LegoSchema](Lego_Schema.PNG)

## Overview

In this assignment, you will load the various tables from the [Rebrickable](https://rebrickable.com/) website and perform analysis on the data.  Rebrickable is a fantastic website that is used to show you which LEGO sets you can build from the sets and parts you already own.  They will even show you which parts you are missing to complete the new set and suggest sets that you could purchase to get those parts.  

## Data

The data has been downloaded directly from their website and includes files for the tables shown in the schema above.  Your first task will be to go through the data and review the schema to understand how the tables all fit together.  There is no data dictionary that I am aware of, but you should be able to understand this data by reviewing the files and schema and visiting their [Download](https://rebrickable.com/downloads/) page and [API](https://rebrickable.com/api/) documentation.  **Note that you must use the files downloaded from Brightspace for this assignment and not download new files from the Rebrickable website.**

## Assignment

In this assignment, you will create a number of functions to complete certain tasks.  This is not needed if you were working on this project on your own, but since we want to see if you can select the data via Pandas/Numpy, we will use those functions in the automatic grading in CodeGrade.  Without this, students could try to look up the answers in the `csv` files instead of wrangling the data with Python/Pandas, which would defeat the purpose of this class.

In addition, we would strongly suggest that you merge data sets as appropriate for the given task.  For example, you might merge the `sets` and `themes` DataFrames to answer one question, and then merge `sets`, `inventories`, `inventory_minifigs`, and `minifigs` to answer another.  This will make your code much easier to write and will be key in working with data across multiple tables.  In fact, I was able to create each function with only one line of code because of the merging of certain data sets.  Also note that you may not use all the tables in this assignment.  This is also an important skill for a Data Scientist to know when to use data and what data is unnecessary.

When complete, save your notebook as `lego.ipynb` and submit to CodeGrade for automatic grading.

## Note

<u>Show Work</u>

Remember that you must show your work.  Students submissions are spot checked manually throughout the term to verify that they are not hard coding the answer from looking only in the file or in CodeGrade's expected output.  If this is seen, the student's answer will be manually marked wrong and their grade will be changed to reflect this. 

For example, if the answer to Q1, the mean of a specific column, is 22:
```
# correct way
Q1 = df['column_name'].mean()

# incorrect way
Q1 = 22 
```

Let's start out by importing some standard imports.

In [147]:
# common imports
import pandas as pd
import numpy as np

# Do not change this option; This allows the CodeGrade auto grading to function correctly
pd.set_option('display.max_columns', None)

## Import Data

Load the various Lego datasets below calling them the same as the name of the table as seen in the schema.  Note that you may or may not have to use all the data sets.  One of the things that all Data Scientists and Analysts have to decide when working on a project is what data is useful for a task and what data is not.  Remember to look at the schema above to see how all the data sets are connected.

The autograder in CodeGrade will have the `csv` files in the same location where the notebook is run.  Because of this, please make sure your local files are in the same location as your notebook.

In [148]:
# import Lego data sets here
# make sure that files are in the same folder as your notebook
inventories = pd.read_csv('inventories.csv')
inventory_sets = pd.read_csv('inventory_sets.csv')
sets = pd.read_csv('sets.csv')
themes = pd.read_csv('themes.csv')
inventory_minifigs = pd.read_csv('inventory_minifigs.csv')
minifigs = pd.read_csv('minifigs.csv')
inventory_parts = pd.read_csv('inventory_parts.csv')
part_categories = pd.read_csv('part_categories.csv')
parts = pd.read_csv('parts.csv')
colors = pd.read_csv('colors.csv')
part_relationships = pd.read_csv('part_relationships.csv')
elements = pd.read_csv('elements.csv')

## Exploration

Take time to look at all of the data sets using `.head()`, `.info()`, `.describe()`, etc to familiarize yourself with the data sets.  As you look at them, notice how the various tables connect together.

In [149]:
### ENTER CODE HERE ###

## Assignment

Answer all the questions below using the available data.  There are multiple ways to get some of the answers.  I would suggest that you attempt to merge some data sets as needed to assist with getting answers.  This will make your tasks a lot easier and your code will be cleaner.

Also, you will be creating functions to get the answers to some of these questions.  This will make it much easier for you to get answers to similar questions in the future and CodeGrade will be checking your functions using different inputs.  

**Q1:** Create a function called `data_count` that takes as input the data set (for example: `sets`) and returns the total number of rows in the data (as an integer).  Nothing else should be returned.

In [150]:
#Create a function called `data_count` 
# that takes as input the data set (for example: `sets`) and returns the total number of rows in the data

def data_count(df):
    return len(df.axes[0])



**Code Check:**  If you call your `data_count` function using the `sets` DataFrame, you should see that there are 18,576 sets represented in the data.

In [151]:
#check the function
data_count(sets)

18576

**Q2:** What is the average number of parts for a Lego set (rounded to the nearest integer)?  Save this as `Q2`.

In [152]:
#find the mean of averge number of parts per set
import math
Q2 = math.ceil(sets['num_parts'].mean())
Q2 


161

**Q3:** What is the median number of parts per set (output as an integer)?  Save this as `Q3`.

In [153]:
#find the median number of parts per set
Q3 = math.floor(sets['num_parts' ].median())
Q3

37

You should see that the median number of parts is much lower than the average number of parts per set.  This tells us that there are some Lego sets with a very high number of parts that brings the average up.

**Q4:** What is the largest number of parts that a set has? Save this as `Q4`.

In [154]:
#find the largest number of parts per set
Q4 = sets['num_parts'].max()
Q4


11695

**Q5:** Given the set with the largest number of parts calculated above, select its row from the `sets` DataFrame.  Save this as `Q5`.  Do not reset the index.

In [207]:
# select the largest numner of parts rows
Q5 = sets.loc[sets["num_parts"] == Q4]
Q5

set_num       name  year  theme_id  num_parts
3867  31203-1  World Map  2021       709      11695

**Q6:** Create a function called `select_set_row` where you pass it the `set_num` (as a string) and the function returns the respective row of the DataFrame from `sets`. Do not reset the index.

In [156]:
#Create a function where you pass it the `set_num` (as a string) and 
# the function returns the respective row of the DataFrame from `sets`
def select_set_row ( set_num ):
    return sets.loc[sets['set_num'] == set_num]



**Q7:** Create a function called `select_set_numparts` that takes as input a specific number of parts (as an integer) and selects all the rows in the `sets` DataFrame that contain those number of parts.  Return the DataFrame sorted by `year` and then by `set_num`.  Do not reset the index.

In [157]:
#Create a function called `select_set_numparts` that
#  takes as input a specific number of parts (as an integer) and 
# selects all the rows in the `sets` DataFrame that contain those number of parts.  
# Return the DataFrame sorted by `year` and then by `set_num`.


def select_set_numparts (num):
    df_with_special_num = sets.loc[sets['num_parts'] == num]
    return df_with_special_num.sort_values(['year', 'set_num'],
              ascending = [True, True])


**Code Check:** Call the `select_set_numparts()` function to select all sets with `0` number of parts.  There should be 3,146 rows returned.

In [158]:
### CODE CHECK ###
select_set_numparts(0)

set_num                                               name  \
11061          700.1-2                     Extra-Large Gift Set (Mursten)   
17563         b55dk-01               Lego Mursten - System i Leg Byggebog   
17564         b56de-01                                    System im Spiel   
11072          700.2-2                                       Gift Package   
11078         700.3A-2                                       Gift Package   
...                ...                                                ...   
15619         854195-1                                    Ernie Key Chain   
15620         854197-1                               Demogorgon Key Chain   
17170  9780744054583-1  DC Comics Super Heroes: Character Encyclopedia...   
17193  9780794447557-1                               City: Color The City   
17200  9780794448172-1           DC Comics Super Heroes: Ready For Action   

       year  theme_id  num_parts  
11061  1953       366          0  
17563  1955       497          0  
17564  1956       497          0  
11072  1957       366          0  
11078  1957       372          0  
...     ...       ...        ...  
15619  2022       503          0  
15620  2022       503          0  
17170  2022       497          0  
17193  2022       497          0  
17200  2022       497          0  

[3146 rows x 5 columns]

**Q8:** 
- The oldest set listed was made in what year?  Save the year as `Q8A` (as an integer).  
- The newest set listed was made in what year? Save the year as `Q8B` (as an integer).

In [159]:
#select the set the oldest year set

Q8A = sets['year'].min()
Q8A

1949

In [160]:
#select the set the newest year set
Q8B = sets['year'].max()
Q8B

2022

**Q9:** 
- Create a function called `select_set_year` that takes as input a year (as an integer) and returns the rows from the `sets` DataFrame that matches that year.  Sort the DataFrame by `set_num`.  Do not reset the index.

In [161]:
#- Create a function called `select_set_year` that 
# takes as input a year (as an integer) and 
# returns the rows from the `sets` DataFrame that matches that year. 
#  Sort the DataFrame by `set_num`.

def select_set_year(the_year):
    return sets.loc[sets['year'] == the_year].sort_values('set_num')


**Q10:** 
- Create a function called `theme_by_year` that takes as input a year (as an integer) and shows the theme ids and theme names (listed in order by theme id) that were in sets that year.  
- The column names must be `id` and `name_themes` (to differentiate between the name of a theme and the name of a set) in that order.  
- The index should be reset and go from `0` to `n-1`.  
- Each theme should only be listed once even if it appeared in more than one set from that year -- duplicate themes should be based on theme id and not name since there are some themes with the same name but with a different id.
- Hint:  It will help if you were to think about merging appropriate DataFrames to help you get this answer.

In [162]:
#change the column name from theme_id to id
sets2 = sets.rename({'theme_id': 'id'}, axis=1)
themes2 = themes.rename({'name': 'name_themes'}, axis=1)
sets2

set_num                             name  year   id  num_parts
0            001-1                            Gears  1965    1         43
1           0011-2                Town Mini-Figures  1979   67         12
2           0011-3       Castle 2 for 1 Bonus Offer  1987  199          0
3           0012-1               Space Mini-Figures  1979  143         12
4           0013-1               Space Mini-Figures  1979  143         12
...            ...                              ...   ...  ...        ...
18571      XWING-1              Mini X-Wing Fighter  2019  158         60
18572      XWING-2                X-Wing Trench Run  2019  158         52
18573  YODACHRON-1  Yoda Chronicles Promotional Set  2013  158        413
18574   YTERRIER-1                Yorkshire Terrier  2018  598          0
18575     ZX8000-1             ZX 8000 LEGO Sneaker  2020  501          0

[18576 rows x 5 columns]

In [163]:
#merge sets and themes

themes_and_sets = pd.merge(sets2, themes2)
themes_and_sets
#themes_and_sets.sort_values(by='year', ascending=True)

set_num                                name  year   id  num_parts  \
0            001-1                               Gears  1965    1         43   
1            002-1      4.5V Samsonite Gears Motor Set  1965    1          3   
2           1030-1      TECHNIC I: Simple Machines Set  1985    1        211   
3           1038-1                  ERBIE the Robo-Car  1985    1        120   
4           1039-1                Manual Control Set 1  1986    1         39   
...            ...                                 ...   ...  ...        ...   
18571  M20-2566-10  Modulex Box - 10x Window 1 x 6 x 6  1963  716         10   
18572  M20-2575-10  Modulex Box - 10x Window 1 x 7 x 5  1963  716         10   
18573  M20-2576-10  Modulex Box - 10x Window 1 x 7 x 6  1963  716         10   
18574  M20-2586-10  Modulex Box - 10x Window 1 x 8 x 6  1963  716         10   
18575  WEETABIX1-1                     Weetabix Castle  1970  414        471   

      name_themes  parent_id  
0         Technic        NaN  
1         Technic        NaN  
2         Technic        NaN  
3         Technic        NaN  
4         Technic        NaN  
...           ...        ...  
18571     Modulex        NaN  
18572     Modulex        NaN  
18573     Modulex        NaN  
18574     Modulex        NaN  
18575      Castle      411.0  

[18576 rows x 7 columns]

In [168]:
# create a fumction 
def theme_by_year(a_year: int):
    result = themes_and_sets.loc[themes_and_sets['year'] == a_year].drop('year', axis=1)
    
    result = result.loc[:, ['id',  'name_themes']].sort_values('id')
    result = result.drop_duplicates(subset=['id'])
    result.reset_index(drop=True, inplace=True)
    

    return result




In [169]:
theme_by_year(1960)

id   name_themes
0  371  Supplemental
1  497         Books
2  513       Classic

**Code Check:** Call the `theme_by_year()` function on the year `1960`.  Your output should look as follows:

| | id | name_themes|
|-|---|---|
|0|371|Supplemental|
|1|497|Books|
|2|513|Classic|

**Optional Practice:** Call the `theme_by_year()` function using your birth year to see the Lego themes that were popular when you were born.

In [170]:
### CODE CHECK / OPTIONAL PRACTICE ###
theme_by_year(1900)

Empty DataFrame
Columns: [id, name_themes]
Index: []

**Q11:** 
- Create a function called `theme_by_name` that takes as input a theme name and outputs all the sets (in `set_num` order) that had that theme.  
- The DataFrame should include the following columns in this order and with these exact names: `set_num`,`name_sets`,`year`,`num_parts`, `theme_id`, and `name_themes`.  
- The index should go from `0` to `n-1`.

In [171]:
#rename the columns
sets3 = sets.rename({'name': 'name_sets'}, axis=1)
sets3

themes1 = themes.rename({'id': 'theme_id', 'name': 'name_themes'}, axis=1)
themes1

#merge sets3 and themes1

df_Q11 = pd.merge(sets3, themes1)
df_Q11 = df_Q11[['set_num','name_sets','year','num_parts', 'theme_id', 'name_themes']]
df_Q11

set_num                           name_sets  year  num_parts  \
0            001-1                               Gears  1965         43   
1            002-1      4.5V Samsonite Gears Motor Set  1965          3   
2           1030-1      TECHNIC I: Simple Machines Set  1985        211   
3           1038-1                  ERBIE the Robo-Car  1985        120   
4           1039-1                Manual Control Set 1  1986         39   
...            ...                                 ...   ...        ...   
18571  M20-2566-10  Modulex Box - 10x Window 1 x 6 x 6  1963         10   
18572  M20-2575-10  Modulex Box - 10x Window 1 x 7 x 5  1963         10   
18573  M20-2576-10  Modulex Box - 10x Window 1 x 7 x 6  1963         10   
18574  M20-2586-10  Modulex Box - 10x Window 1 x 8 x 6  1963         10   
18575  WEETABIX1-1                     Weetabix Castle  1970        471   

       theme_id name_themes  
0             1     Technic  
1             1     Technic  
2             1     Technic  
3             1     Technic  
4             1     Technic  
...         ...         ...  
18571       716     Modulex  
18572       716     Modulex  
18573       716     Modulex  
18574       716     Modulex  
18575       414      Castle  

[18576 rows x 6 columns]

In [172]:
 # - Create a function called `theme_by_name` 
 # that takes as input a theme name and
 #  outputs all the sets (in `set_num` order) that had that theme.  
#- The DataFrame should include the following columns in this order and with these exact names: 
def theme_by_name (t_name):
      result  = df_Q11.loc[df_Q11["name_themes"] == t_name].sort_values('set_num').reset_index(drop=True)
      return result







**Code Check:** Call the `theme_by_name()` function using the theme name `Gear`.  You should return 1,904 sets that match that theme name.

In [173]:
### CODE CHECK ###
theme_by_name('Gear')

set_num                                          name_sets  \
0              11138-1                            Bionicle Barraki Wallet   
1       119645111436-1                Classic Space Astronaut Plush (Red)   
2       119645111504-1               Classic Space Astronaut Plush (Blue)   
3       119645111740-1              Classic Space Astronaut Plush (White)   
4       119645111818-1               Classic Space Astronaut Plush (Pink)   
...                ...                                                ...   
1899           WATCH-3                   Watch System Voyager (UFO Space)   
1900         WIIGAME-1                              Bionicle Heroes - Wii   
1901         WIIGAME-2  Indiana Jones: The Original Adventures Video G...   
1902  WOODENELEPHANT-1  Original Wooden Elephant Pull-Along Toy and Print   
1903          ZX8000-1                               ZX 8000 LEGO Sneaker   

      year  num_parts  theme_id name_themes  
0     2008          0       501        Gear  
1     2021          0       501        Gear  
2     2021          0       501        Gear  
3     2021          0       501        Gear  
4     2021          0       501        Gear  
...    ...        ...       ...         ...  
1899  1997         29       501        Gear  
1900  2007          0       501        Gear  
1901  2008          0       501        Gear  
1902  2021          0       501        Gear  
1903  2020          0       501        Gear  

[1904 rows x 6 columns]

**Q12:** Create a function called `theme_by_setnum` that takes as input the set_num (as a string) and returns the respective theme name as a string.

In [174]:
#Create a function called `theme_by_setnum` that 
# takes as input the set_num (as a string) and 
# returns the respective theme name as a string

def theme_by_setnum(a: str):
    df_whole_row = df_Q11.loc[df_Q11["set_num"] == a] 
    result = df_whole_row ['name_themes'].values[0]
    return result


**Code Check:** Visit the [Sets page](https://rebrickable.com/sets/) from the Rebrickable website and find a set that you like.  Call your `theme_by_setnum()` function on that set and make sure the correct theme name is output.

In [175]:
### CODE CHECK ###
theme_by_setnum('21036-1')


'Architecture'

**Millennium Falcon Prep Exercise:** Find all the sets named 'Millennium Falcon'. Of these sets, find the set with the largest number of parts.  You will use this information for later questions.

In [176]:
#Find all the sets named 'Millennium Falcon'.
Millennium_Falcon_set =  themes_and_sets.loc[themes_and_sets["name"] == 'Millennium Falcon'].sort_values('num_parts').reset_index(drop=True)
Millennium_Falcon_set


set_num               name  year   id  num_parts  \
0     TRUSWMF-2  Millennium Falcon  2018  158         24   
1      911949-1  Millennium Falcon  2019  158         32   
2      911607-1  Millennium Falcon  2016  158         42   
3   TRUFALCON-1  Millennium Falcon  2016  158         44   
4        4488-1  Millennium Falcon  2003  158         87   
5       75030-1  Millennium Falcon  2014  158         94   
6        7190-1  Millennium Falcon  2000  158        671   
7        7965-1  Millennium Falcon  2011  158       1254   
8       75105-1  Millennium Falcon  2015  158       1330   
9       75257-1  Millennium Falcon  2019  158       1351   
10      10179-1  Millennium Falcon  2007  171       5197   
11      75192-1  Millennium Falcon  2017  171       7541   

                  name_themes  parent_id  
0                   Star Wars        NaN  
1                   Star Wars        NaN  
2                   Star Wars        NaN  
3                   Star Wars        NaN  
4                   Star Wars        NaN  
5                   Star Wars        NaN  
6                   Star Wars        NaN  
7                   Star Wars        NaN  
8                   Star Wars        NaN  
9                   Star Wars        NaN  
10  Ultimate Collector Series      158.0  
11  Ultimate Collector Series      158.0

In [214]:
#find the set with the largest number of parts.
MAx_num_parts = Millennium_Falcon_set['num_parts'].max()
MAx_num_parts



7541

In [220]:


Millennium_Falcon_set_theme_id =Millennium_Falcon_set.loc[Millennium_Falcon_set["num_parts"] == MAx_num_parts]
Millennium_Falcon_set_theme_id
Millennium_Falcon_set_theme_id['id']

11    171
Name: id, dtype: int64

**Q13:** 
- How many minifigs are contained in the set from the `Millennium Falcon Prep Exercise`?  Save this as `Q13` (as an integer).  
- Note that there are sometimes multiple versions of a set as shown in the `inventories` table.  This will not matter for this question as there is only one version for the respective set but this should be noted for future questions.
- Remember to show your work or your answer will manually get marked as incorrect.

In [178]:
#merge and rename dataframes
#df_Q13_1 = pd.merge(sets, inventories, on = 'set_num')
#df_Q13_1
#df_Q13_11 = df_Q13_1.rename({'id': 'inventory_id'}, axis=1)
#df_Q13_11
#df_Q13_2 = pd.merge(inventory_minifigs, minifigs, on = 'fig_num')
#df_Q13_2
#df_Q13_3 = pd.merge(df_Q13_11, df_Q13_2, on = 'inventory_id')

In [179]:
#Millennium_Falcon_set =  df_Q13_3.loc[df_Q13_3["name_x"] == 'Millennium Falcon'].sort_values('set_num').reset_index(drop=True)
#Millennium_Falcon_set = Millennium_Falcon_set.drop_duplicates(['set_num'])
#Millennium_Falcon_set

# all_set_numbers = set(df_Q13_3['set_num'].values)
# small_df_set_numbers = set(Millennium_Falcon_set['set_num'].values)
# common_set_numbers = all_set_numbers.intersection(small_df_set_numbers)
# common_set_numbers


In [180]:
#Rename columns
inventories2 = inventories.rename({'id': 'inventory_id'}, axis=1)
inventories2 

#Merge files-themes_and_sets and inventories2
df_Q13_1 = pd.merge(themes_and_sets, inventories2)
df_Q13_1

#Merge files-   df_Q13_1 and inventory_minifigs
df_Q13_2 = pd.merge(df_Q13_1, inventory_minifigs)
df_Q13_2

#Rename minifigs file columns
minifigs2 = minifigs.rename({'name': 'fig_name', 'num_parts': 'fig_num_parts'}, axis=1)
minifigs2 

#Merge files-   df_Q13_2 and minifigs2 
df_Q13 = pd.merge(df_Q13_2, minifigs2)



In [181]:
#How many minifigs are contained in the set from the `Millennium Falcon Prep Exercise`?  Save this as `Q13`
Q13_1 =  df_Q13.loc[df_Q13["name"] == 'Millennium Falcon'].sort_values('fig_num').reset_index(drop=True)
Q13_1 = Q13_1.drop_duplicates(subset=['set_num'])
Q13 = len(Q13_1.index)
Q13

7

**Q14:** 
- Create a function called `get_minifigs` that takes as input a set number (as a string) and a version number (as an integer) and outputs all minifigs from that set / version combination.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `set_num`, `fig_num`, `fig_name`, `fig_num_parts`.  
- The index should go from 0 to n-1.

In [182]:
#select data I need
df_Q14 = df_Q13[['set_num', 'fig_num', 'fig_name', 'fig_num_parts', 'version']]



In [183]:
#- Create a function called `get_minifigs` that 
# takes as input a set number (as a string) and a version number (as an integer) and 
# outputs all minifigs from that set / version combination.  
# `set_num`, `fig_num`, `fig_name`, `fig_num_parts`.  
def get_minifigs(set_number : str, version_number):
    result = df_Q14.loc[(df_Q14['set_num'] == set_number) & (df_Q14['version'] == version_number)].drop('version', axis=1).sort_values('fig_num').reset_index(drop=True)
    return result



set_num     fig_num                                           fig_name  \
0   10217-1  fig-002229                   Skeleton (Standard) (3626b Head)   
1   10217-1  fig-006075                 Rubeus Hagrid, Dark Brown Overcoat   
2   10217-1  fig-006084  Lucius Malfoy, Black Robes and Cape, Death Eat...   
3   10217-1  fig-006130  Harry Potter, Open Dark Blue Jacket over White...   
4   10217-1  fig-006138                                    Fenrir Greyback   
5   10217-1  fig-006162  Ron Weasley, Red Tartan Sweater Vest, Reddish ...   
6   10217-1  fig-006163          Fred / George Weasley in Dark Orange Suit   
7   10217-1  fig-006164                                 Garrick Ollivander   
8   10217-1  fig-006165    Goblin, Black Jacket and Light Bluish Gray Legs   
9   10217-1  fig-006166        Goblin, Black Jacket and Reddish Brown Legs   
10  10217-1  fig-006167  Hermione Granger, Light Bluish Gray Sweater wi...   

    fig_num_parts  
0               6  
1               3  
2               5  
3               4  
4               4  
5               4  
6               4  
7               4  
8               3  
9               3  
10              4

**Q15:** Using the above function, output information about the minifigs from the set discussed in the `Millennium Falcon Prep Exercise`.  Save this output as `Q15`.

In [206]:
check the version 
M =  df_Q13.loc[(df_Q13["name"] == 'Millennium Falcon') & (df_Q13["num_parts"] == 7541)]
M

set_num               name  year   id  num_parts  \
1791  75192-1  Millennium Falcon  2017  171       7541   
2564  75192-1  Millennium Falcon  2017  171       7541   
2584  75192-1  Millennium Falcon  2017  171       7541   
2635  75192-1  Millennium Falcon  2017  171       7541   
2640  75192-1  Millennium Falcon  2017  171       7541   
2650  75192-1  Millennium Falcon  2017  171       7541   
2651  75192-1  Millennium Falcon  2017  171       7541   
2652  75192-1  Millennium Falcon  2017  171       7541   

                    name_themes  parent_id  inventory_id  version     fig_num  \
1791  Ultimate Collector Series      158.0         19670        1  fig-002057   
2564  Ultimate Collector Series      158.0         19670        1  fig-001714   
2584  Ultimate Collector Series      158.0         19670        1  fig-002514   
2635  Ultimate Collector Series      158.0         19670        1  fig-001810   
2640  Ultimate Collector Series      158.0         19670        1  fig-002544   
2650  Ultimate Collector Series      158.0         19670        1  fig-002549   
2651  Ultimate Collector Series      158.0         19670        1  fig-002550   
2652  Ultimate Collector Series      158.0         19670        1  fig-002551   

      quantity                                           fig_name  \
1791         1                                 Rey, Dark Tan Robe   
2564         1                              Chewbacca, Dark Brown   
2584         1    C-3PO, Pearl Gold, Colorful Wires, Printed Legs   
2635         1                             Finn, Black Undershirt   
2640         1                                               BB-8   
2650         1  Princess Leia, Hoth Outfit, White Jacket, Prin...   
2651         1  Han Solo, Dark Blue Jacket, Dark Brown Legs, P...   
2652         1                               Han Solo, Old, Angry   

      fig_num_parts  
1791              4  
2564              3  
2584              3  
2635              4  
2640              2  
2650              4  
2651              4  
2652              4

In [208]:
#check 
Q15 = get_minifigs('75192-1', 1)
Q15

set_num     fig_num                                           fig_name  \
0  75192-1  fig-001714                              Chewbacca, Dark Brown   
1  75192-1  fig-001810                             Finn, Black Undershirt   
2  75192-1  fig-002057                                 Rey, Dark Tan Robe   
3  75192-1  fig-002514    C-3PO, Pearl Gold, Colorful Wires, Printed Legs   
4  75192-1  fig-002544                                               BB-8   
5  75192-1  fig-002549  Princess Leia, Hoth Outfit, White Jacket, Prin...   
6  75192-1  fig-002550  Han Solo, Dark Blue Jacket, Dark Brown Legs, P...   
7  75192-1  fig-002551                               Han Solo, Old, Angry   

   fig_num_parts  
0              3  
1              4  
2              4  
3              3  
4              2  
5              4  
6              4  
7              4

**Code Check:** Make sure that your `get_minifigs()` function works with both version 1 and version 2 of set 10217-1, Harry Potter's Diagon Alley.  Both versions should have 11 minifigs in the DataFrame.

In [209]:
### CODE CHECK ###
get_minifigs('10217-1', 1)

set_num     fig_num                                           fig_name  \
0   10217-1  fig-002229                   Skeleton (Standard) (3626b Head)   
1   10217-1  fig-006075                 Rubeus Hagrid, Dark Brown Overcoat   
2   10217-1  fig-006084  Lucius Malfoy, Black Robes and Cape, Death Eat...   
3   10217-1  fig-006130  Harry Potter, Open Dark Blue Jacket over White...   
4   10217-1  fig-006138                                    Fenrir Greyback   
5   10217-1  fig-006162  Ron Weasley, Red Tartan Sweater Vest, Reddish ...   
6   10217-1  fig-006163          Fred / George Weasley in Dark Orange Suit   
7   10217-1  fig-006164                                 Garrick Ollivander   
8   10217-1  fig-006165    Goblin, Black Jacket and Light Bluish Gray Legs   
9   10217-1  fig-006166        Goblin, Black Jacket and Reddish Brown Legs   
10  10217-1  fig-006167  Hermione Granger, Light Bluish Gray Sweater wi...   

    fig_num_parts  
0               6  
1               3  
2               5  
3               4  
4               4  
5               4  
6               4  
7               4  
8               3  
9               3  
10              4

**Q16:** 
- Create a function called `minifigs_from_themes` that takes as input a theme id (as an integer) and outputs all the minifigs contained in sets with that theme.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `fig_num`, `fig_name`. 
- The index should go from 0 to n-1.  
- Exact duplicates should be removed so that only unique figs are included.

In [186]:
#df_Q16_1 = df_Q13.rename({'id': 'theme_id'}, axis=1)
df_Q16 = df_Q13[['fig_num', 'fig_name', 'id']]



In [187]:
#- Create a function called `minifigs_from_themes` that 
# takes as input a theme id (as an integer) and 
# outputs all the minifigs contained in sets with that theme.  
#`fig_num`, `fig_name`. 
def minifigs_from_themes(theme_id):
    df_new = df_Q16.loc[(df_Q16['id'] == theme_id) ].drop('id', axis=1).sort_values('fig_num')
    result = df_new.drop_duplicates(subset=['fig_num']).reset_index(drop=True)
    return result


**Code Check:** Call the `minifigs_from_themes()` function using the theme number of the set discussed in the `Millennium Falcon Prep Exercise`.  There should be 149 minifigs output for that respective theme.

In [221]:
#code check   Theme_id is 171
minifigs_from_themes(Millennium_Falcon_set_theme_id)

C:\Users\Lucy Wu\AppData\Local\Temp\ipykernel_17604\3425790052.py:6: FutureWarning: Automatic reindexing on DataFrame vs Series comparisons is deprecated and will raise ValueError in a future version. Do `left, right = left.align(right, axis=1, copy=False)` before e.g. `left == right`
  df_new = df_Q16.loc[(df_Q16['id'] == theme_id) ].drop('id', axis=1).sort_values('fig_num')


ValueError: Cannot index with multidimensional key

**Q17:** 
- Create a function called `sets_from_minifig` that takes as input a minifig number (as a string) and returns a list of sets a minifig has appeared in.  
- Return a DataFrame (sorted by `set_num` and then by `version`) with the following columns: `set_num`, `version`,`name_sets`, `fig_num`, `fig_name`.  
- The index should go from 0 to n-1.

In [189]:
#make the dataframe we need
df_Q17_1 = df_Q13.rename({'name': 'name_sets'}, axis=1)
df_Q17 = df_Q17_1[['set_num', 'version','name_sets', 'fig_num', 'fig_name']]


In [190]:
#- Create a function called `sets_from_minifig` that 
# takes as input a minifig number (as a string) and 
# returns a list of sets a minifig has appeared in.  
####- Return a DataFrame (sorted by `set_num` and then by `version`) with the following columns: `set_num`, `version`,`name_sets`, `fig_num`, `fig_name`.  

def sets_from_minifig(minifig_number):
    df_new = df_Q17.loc[(df_Q17['fig_num'] == minifig_number)]
    result = df_new.reset_index(drop=True)
    return result



**Code Check:** Call your `sets_from_minifig()` function on one of the Lego Friends minifigs, Andrea, who has figure number: `fig-001039`.  Your function should return a DataFrame that contains the following sets:

- Andrea on the Beach
- Birthday Party
- City Park Cafe

In [191]:
### CODE CHECK ###
sets_from_minifig('fig-001039')



set_num  version            name_sets     fig_num  \
0  30100-1        1  Andrea on the Beach  fig-001039   
1  30107-1        1       Birthday Party  fig-001039   
2   3061-1        1       City Park Café  fig-001039   

                                            fig_name  
0  Andrea - Bright Light Orange Top, Light Aqua S...  
1  Andrea - Bright Light Orange Top, Light Aqua S...  
2  Andrea - Bright Light Orange Top, Light Aqua S...

**Q18:** 

*Note:  Before you begin working on this question, please note that CodeGrade has a 1GB RAM limit.  If you see an `Exit -9` code when submitting your assignment, it is probably due to this RAM limit and previous students had an issue with this question causing too much memory usage.  The biggest thing to think about is to only use the tables that you need to answer for this question.  For instance, you do not need any of the minifigs data since this is not needed for the output and the merging of this data could cause the number of rows to "explode" in terms of the total count.  I understand that this can be annoying that we have to work around CodeGrade's limitations, but it is good practice in memory management.*

*You should only need to merge the following datasets for this answer so as to not go above the RAM limit: `themes`, `sets`, `inventories`, `inventory_parts`, `parts`, and `colors`*

- Create a function called `set_parts` that takes as input the set number (as a string) and version number (as an integer) and outputs a DataFrame with each part in the set.  
- The DataFrame should have the following columns/column names: `set_num`, `version`, `name_sets`, `year`, `name_themes`, `part_num`, `name_part`, `name_color`,`quantity`,`is_spare`.  
- The Data should be sorted by `part_num` and then `name_color`.  
- The index should go from 0 to n-1.

In [192]:
#rename column's name
themes_re = themes.rename({'id': 'theme_id', 'name': 'name_themes'}, axis=1)
sets_re = sets.rename({'name': 'name_sets'}, axis=1)
inventories_re = inventories.rename({'id': 'inventory_id'}, axis=1)
parts_re = parts.rename({'name': 'name_part'}, axis=1)
colors_re = colors.rename({'name': 'name_color', 'id': 'color_id'}, axis=1)

In [193]:
#merge the files

df1 = pd.merge(sets_re, themes_re, on= 'theme_id')
#`themes`, `sets`, `inventories`
df2 = pd.merge(df1, inventories_re)
# , `inventory_parts`, 
df3 = pd.merge(df2, inventory_parts, on = 'inventory_id')
#`parts`
df4 = pd.merge(df3, parts_re, on = 'part_num')
#colors`
df_Q18_1 = pd.merge(df4, colors_re, on = 'color_id')

df_Q18 = df_Q18_1[['set_num', 'version','name_sets', 'year', 
                    'name_themes','part_num', 'name_part', 
                    'name_color', 'quantity','is_spare']]


In [194]:
#create Q18 function
def set_parts(set_number, version_number):
    result = df_Q18.loc[(df_Q18['set_num'] == set_number) & (df_Q18['version'] == version_number)].sort_values('part_num').reset_index(drop=True)
    return result


**Code Check**: Using the `set_parts()` function, output a list of parts for the set discussed in the `Millennium Falcon Prep Exercise`. There should be a total of 730 rows in the DataFrame, which includes 676 parts and 54 spare parts. You can check the respective page on the rebrickable website to verify these results.

In [196]:
### CODE CHECK ###
set_parts('30100-1',1)

set_num  version            name_sets  year name_themes part_num  \
0   30100-1        1  Andrea on the Beach  2012     Friends     3032   
1   30100-1        1  Andrea on the Beach  2012     Friends    33121   
2   30100-1        1  Andrea on the Beach  2012     Friends     3837   
3   30100-1        1  Andrea on the Beach  2012     Friends    4032a   
4   30100-1        1  Andrea on the Beach  2012     Friends    4495b   
5   30100-1        1  Andrea on the Beach  2012     Friends    48245   
6   30100-1        1  Andrea on the Beach  2012     Friends    59275   
7   30100-1        1  Andrea on the Beach  2012     Friends    59275   
8   30100-1        1  Andrea on the Beach  2012     Friends    59900   
9   30100-1        1  Andrea on the Beach  2012     Friends     6141   
10  30100-1        1  Andrea on the Beach  2012     Friends     6141   
11  30100-1        1  Andrea on the Beach  2012     Friends     6141   
12  30100-1        1  Andrea on the Beach  2012     Friends     6141   
13  30100-1        1  Andrea on the Beach  2012     Friends     6141   
14  30100-1        1  Andrea on the Beach  2012     Friends     6141   
15  30100-1        1  Andrea on the Beach  2012     Friends    64644   
16  30100-1        1  Andrea on the Beach  2012     Friends    64648   

                                            name_part           name_color  \
0                                         Plate 4 x 6                  Tan   
1                                        Animal, Crab  Bright Light Orange   
2                   Equipment Shovel [Round Stem End]     Dark Bluish Gray   
3   Plate Round 2 x 2 with Axle Hole Type 1 (+ Ope...                  Tan   
4                  Flag 4 x 1 Wave [First Wave Right]            Dark Pink   
5   Bucket with Handle Holes and 5 Bottom Holes - ...            Dark Pink   
6                     Minifig Footwear Flipper [Thin]                 Lime   
7                     Minifig Footwear Flipper [Thin]                 Lime   
8                             Cone 1 x 1 [Top Groove]                  Tan   
9                   Plate Round 1 x 1 with Solid Stud                  Tan   
10                  Plate Round 1 x 1 with Solid Stud      Trans-Dark Blue   
11                  Plate Round 1 x 1 with Solid Stud     Trans-Light Blue   
12                  Plate Round 1 x 1 with Solid Stud     Trans-Light Blue   
13                  Plate Round 1 x 1 with Solid Stud                  Tan   
14                  Plate Round 1 x 1 with Solid Stud      Trans-Dark Blue   
15                                Equipment Telescope           Pearl Gold   
16                     Animal, Fish [Anti-Stud Mouth]               Orange   

    quantity is_spare  
0          1        f  
1          1        f  
2          1        f  
3          1        f  
4          1        f  
5          1        f  
6          2        f  
7          2        t  
8          5        f  
9          2        t  
10         1        t  
11         1        t  
12         3        f  
13         3        f  
14         3        f  
15         1        f  
16         1        f

## Optional Exercises
These exercises are optional and are not graded by CodeGrade.  They can be used as extra optional practice.

- What is the median year for the sets data?  Given the oldest and newest year in the data, what does this median value tell you about whether the number of sets have increased or decreased over the years?
- What minifig has the most number of parts?
- What minifig shows up in the most number of sets?
- When was the first minifig included in a set?
- What set has the most number of minifigs
- What part is used the most for all sets? 
- What color is used the most for all parts?
- Get a list of all unique color names, sort by color name, and output as a list. 
- How many total transparent colors are in the colors DataFrame?
- Create a function called get_part_colors that takes as input a part number (as a string) and returns a DataFrame of all colors a part has appeared in.
- What part / color combination is used the most for all sets?
- What part / color combination is used the least for all sets?
- Explore some of your favorite themes and which colors are used the most for those themes.
- Plot a line graph of the average number of parts per set per year.
